Source/MSBuildTools.Unity/Packages/com.microsoft.msbuildforunity/Editor/ProjectGenerator/Scripts/MSBuildTools.cs

```diff
@@ -275,58 +275,86 @@ private static void RefreshGeneratedOutput(bool forceGenerateEverything, bool fo
+            bool doesCurrentVersionTokenFileExist = tokenVerison != null && tokenVerison == MSBuildForUnityVersion;
+
+            // We perform the regeneration of complete or partial pass in the following cases:
+            // - forceGenerateEverything is true (we are told to)
+            // - buildTargetOrFrameworkChanged is true (target framework changed)
+            // - doesCurrentVersionTokenFileExist is false (version changed, or editor just opened)
 
-            // We regenerate everything if:
-            // - We are forced to
             // - AutoGenerateEnabled and token file doesn't exist or shouldClean is true
-            bool regenerateEverything = forceGenerateEverything || (Config.AutoGenerateEnabled && (!doesTokenFileExist || shouldClean));
+            bool performRegeneration = forceGenerateEverything || buildTargetOrFrameworkChanged || !doesCurrentVersionTokenFileExist;
 
-            if (regenerateEverything || unityProjectInfo == null)
+            if (performRegeneration || unityProjectInfo == null)
             {
                 // Create the project info only if it's null or we need to regenerate
-                unityProjectInfo = new UnityProjectInfo(Debug.unityLogger, SupportedBuildTargets, Config, Config.AutoGenerateEnabled || forceCompleteGeneration);
+                unityProjectInfo = new UnityProjectInfo(Debug.unityLogger, SupportedBuildTargets, Config, Config.FullGenerationEnabled || forceCompleteGeneration);
             }
 
-            // We regenerate the common "directory" props file under the following conditions:
-            // - Token file doesn't exist which means editor was just started
-            // - EditorPrefs currentBuildTarget or targetFramework is different from current ones (same as for executing a clean)
-            if (shouldClean || !doesTokenFileExist)
+            if (performRegeneration)
             {
-                MSBuildUnityProjectExporter.ExportCommonPropsFile(Exporter, unityProjectInfo.CurrentPlayerPlatform);
+                // If we are forced complete, then we regenerate, otherwise perform the one that is selected
+                RegenerateEverything(unityProjectInfo, Config.FullGenerationEnabled || forceCompleteGeneration);
             }
-
-            if (regenerateEverything)
+            
+            if (!doesCurrentVersionTokenFileExist)
             {
-                RegenerateEverything(unityProjectInfo, Config.AutoGenerateEnabled || forceCompleteGeneration);
-            }
+                foreach (string tokenFile in Directory.GetFiles(Path.Combine(Utilities.ProjectPath, "Temp"), "*_token.msb4u", SearchOption.TopDirectoryOnly))
+                {
+                    File.Delete(tokenFile);
+                }
 
-            if (!doesTokenFileExist)
-            {
-                File.Create(TokenFilePath).Dispose();
+                File.Create(Path.Combine(Utilities.ProjectPath, "Temp", $"{MSBuildForUnityVersion.ToString(3)}_token.msb4u"))
```

Comment:
>(3) [](start = 106, length = 3)

ToString(3) still has a vestige of the hardcoded nature of this version string (same with the token parsing thing below), but I think in this case it's less worrisome because it's contained within a single place (not copied across files)